### DISCUSSION TITLE
too early or late step call

### DISCUSSION THREAD

**Discussion by NDR008** [trackmania-rl/tmrl#67] [Q&A]
What happens when a call to step() of rtgym is called too early? Say after 5ms when a timestep of 50ms was set?

**Comment by yannbouteiller** [comment]
The call blocks untill an observation is captured. This is how rtgym synchronizes timesteps, and it expects you to call step before the end of the timestep (within some allowed elasticity) otherwise it times out and throws a warning.

**Comment by NDR008** [comment]
May need to come back to this question. After I process it better. :)

**Reply by yannbouteiller** [reply]
Well, the flow is approximately this:

- You send an action to `rtgym` and retrieve an observation from `step()`. At this point `rtgym` starts applying this action.

- Meanwhile, you compute a new action from the retrieved observation, and call `step()` again to send this action to `rtgym` when you are done (before the current time-step ends).

- `rtgym` waits untill the end of the current time-step to return with a new observation, and starts applying your computed action.
